### PR TITLE
Clickhouse: new database engine (experimental)

### PR DESCRIPTION
**Pull request by pnax**
## Purpose

[Clickhouse](https://clickhouse.com/) is a vertical database engine that appears to be quite efficient when it comes to run Zonemaster with millions of domains

## Context

* Clickhouse is currently used internally at Afnic to monthly store Zonemaster reslults on the .fr zone (4 millions domains) 

## Changes

* new Clickhouse database engine
* uses the MySQL DBI to connect to the Clickhouse server

### Setup a Clickhouse server

* [install Clickhouse](https://clickhouse.com/docs/en/install)
* Generate a [double SHA1 password](https://clickhouse.com/docs/en/operations/settings/settings-users#password_double_sha1_hex) for the database user
   ```
   PASSWORD=$(base64 < /dev/urandom | head -c8); echo "$PASSWORD"; echo -n "$PASSWORD" | sha1sum | tr -d '-' | xxd -r -p | sha1sum | tr -d '-'
   ```
* Setup a `zonemaster` database and a `zonemaster` user:
   ```
   $ clickhouse-client
   :) CREATE DATABASE zonemaster;
   :) CREATE USER zonemaster IDENTIFIED WITH double_sha1_hash BY '<DOUBLE_SHA1>';
   :) GRANT CREATE TABLE, DROP TABLE, SELECT, INSERT, ALTER UPDATE ON zonemaster.* TO zonemaster;
   ```

### Configure and use Zonemaster-Backend

Adapt the `share/backend_config.ini` file and use `zmb` or `zmtest`.
You could check any results by running some SQL commands with `clickhouse-client`.
```
$ clickhouse-client
:) USE zonemaster;
:) SELECT * FROM test_results;
```

### Run the unit tests

Unit test should work when using Clickhouse.
```
TARGET=Clickhouse prove -l t/
```

### CODE REVIEW

File: docs/Configuration.md [hunk 1]
```diff
@@ -216,6 +222,56 @@ An absolute path.
...
+### port
+
+The port the Clickhouse server is listening on.
+Default value: `9004`.
+
+> TODO update following paragraph
+
+If [MYSQL.host] is set to `localhost` (but neither `127.0.0.1` nor `::1`),
+then the value of the [MYSQL.port] property is discarded as the driver
+connects using a UNIX socket (see the [DBD::mysql documentation]).
```
[the remark this answers is not in the document]
Comment (pnax) [reply docs/Configuration.md:246]: Not sure. It looks like an enthusiastic copy/paste session.

**Comment by matsduf** [comment]
Some installation instructions are needed.

File: lib/Zonemaster/Backend/Config.pm [hunk 2]
```diff
@@ -413,15 +442,15 @@ Returns a normalized string based on the supported databases.
...
+Dies if the value is not one of SQLite, PostgreSQL, MySQL or Clickhouse.
 
 =cut
 
 sub check_db {
     my ( $self, $db ) = @_;
 
     $db = untaint_engine_type( $db )    #
-      // die "Unknown database '$db', should be one of SQLite, MySQL or PostgreSQL\n";
+      // die "Unknown database '$db', should be one of SQLite, MySQL, PostgreSQL or Clikchouse\n";
```
Comment (marc-vanderwal) [inline lib/Zonemaster/Backend/Config.pm:453]:
```suggestion
      // die "Unknown database '$db', should be one of SQLite, MySQL, PostgreSQL or Clickhouse\n";
```

File: lib/Zonemaster/Backend/DB.pm [hunk 3]
```diff
@@ -511,16 +511,15 @@ sub get_test_history {
     my @results;
     my $query = q[
         SELECT
-            (SELECT count(*) FROM result_entries WHERE result_entries.hash_id = test_results.hash_id AND level = ?) AS nb_critical,
-            (SELECT count(*) FROM result_entries WHERE result_entries.hash_id = test_results.hash_id AND level = ?) AS nb_error,
-            (SELECT count(*) FROM result_entries WHERE result_entries.hash_id = test_results.hash_id AND level = ?) AS nb_warning,
-            id,
+            (SELECT count(*) FROM result_entries JOIN test_results ON result_entries.hash_id = test_results.hash_id AND level = ?) AS nb_critical,
+            (SELECT count(*) FROM result_entries JOIN test_results ON result_entries.hash_id = test_results.hash_id AND level = ?) AS nb_error,
+            (SELECT count(*) FROM result_entries JOIN test_results ON result_entries.hash_id = test_results.hash_id AND level = ?) AS nb_warning,
```
Comment (matsduf) [inline lib/Zonemaster/Backend/DB.pm:516]: Are these changes related to the addition of support of Clickhouse engine?

There are other changes in this module where it is not obvious that they are related to the Clickhouse change.

**Review by matsduf — COMMENTED** [review]
This PR is still marked as "draft". Is it really ready for installation?

There are other changes in the code that seems to be unrelated to Clickhouse.

* I find no installation instructions. [Zonemaster-Backend installation document](https://github.com/zonemaster/zonemaster/blob/master/docs/public/installation/zonemaster-backend.md) needs to be updated.
* Also the [Zonemaster-Backend configuration document](https://github.com/zonemaster/zonemaster/blob/master/docs/public/configuration/backend.md) needs to be updated.

File: lib/Zonemaster/Backend/DB.pm (hunk 3 above)
Comment (matsduf) [reply lib/Zonemaster/Backend/DB.pm]: Thanks. That makes it easier to review. I think it looks fine, but Travis does not. Some kind of documentation is needed (installation and configuration).

**Comment by marc-vanderwal** [comment]
# Release testing report – Success, no issues

Rocky Linux 9.3/Clickhouse

Merged `pnax/clickhouse` in a detached head. Followed the installation instructions as updated by zonemaster#1228 in order to get a Clickhouse server running on the same host as zonemaster-backend. After running the “smoke test”, the `test_results` table contains data, as expected.

For the unit tests, the procedure was incomplete. Unit tests needed a `zonemaster_testing` database, a `zonemaster_testing` user and enough privileges for the aforementioned user to touch the database. With those prerequisites out of the way, unit tests pass.

**Review by matsduf — APPROVED** [review]
Assuming it passes CI.

**Comment by marc-vanderwal** [comment]
I’m currently testing #1121 with the experimental Clickhouse database engine still enabled and it seems that I’m hitting issues related to the way Clickhouse implements ALTER TABLE … UPDATE that don’t make me feel comfortable at all with using a Clickhouse database backend in production.

For example, I get warnings like the following:
```
Argument "860b459647d9be55" isn't numeric in numeric eq (==) at /usr/local/share/perl5/5.32/Zonemaster/Backend/DB.pm line 362.
```

That line 362 is the equality comparison of `$progress` with 0 in the following subroutine:
```perl startline=346
sub test_state {
    my ( $self, $test_id ) = @_;

    my ( $progress ) = $self->dbh->selectrow_array(
        q[
            SELECT progress
            FROM test_results
            WHERE hash_id = ?
        ],
        undef,
        $test_id,
    );
    if ( !defined $progress ) {
        die Zonemaster::Backend::Error::Internal->new( reason => 'job not found' );
    }

    if ( $progress == 0 ) {
        return $TEST_WAITING;
    }
    elsif ( 0 < $progress && $progress < 100 ) {
        return $TEST_RUNNING;
    }
    elsif ( $progress == 100 ) {
        return $TEST_COMPLETED;
    }
    else {
        die Zonemaster::Backend::Error::Internal->new( reason => 'state could not be determined' );
    }
}
```

But these “argument isn’t numeric” errors should in theory never happen, because the code fetches exactly one row and column and the datum is, according to the database schema, an integer. How do we end up with a string?

Grepping `860b459647d9be55` across both log files shows something even more concerning: the thing that’s being compared to a number is actually the ID of a test being run by the *other* of the two test agents on the machine. In this example, if I order the matches by timestamp, this happens:
```
zm-testagent-2.log:2023-12-13T08:59:51Z [2296] [WARNING] [main] Argument "860b459647d9be55" isn't numeric in numeric eq (==) at /usr/local/share/perl5/5.32/Zonemaster/Backend/DB.pm line 362.
zm-testagent.log:2023-12-13T08:59:53Z [2152] [INFO] [main] Test found: 860b459647d9be55
zm-testagent-2.log:2023-12-13T08:59:54Z [2296] [WARNING] [main] Argument "860b459647d9be55" isn't numeric in numeric eq (==) at /usr/local/share/perl5/5.32/Zonemaster/Backend/DB.pm line 362.
zm-testagent.log:2023-12-13T09:00:21Z [2812] [INFO] [main] Test starting: 860b459647d9be55
zm-testagent.log:2023-12-13T09:05:53Z [2812] [INFO] [main] Test completed: 860b459647d9be55
zm-testagent.log:2023-12-13T09:05:54Z [2152] [NOTICE] [Zonemaster::Backend::Config] Worker process (pid 2812, testid 860b459647d9be55): Terminated with exit code 0
```

Due to this, many tests either crash with “state could not be determined” or “illegal transition to COMPLETED”.

It seems that Clickhouse doesn’t implement proper transaction isolation when updating a table. Yet, currently, the code does many updates to `test_results` that, without proper isolation, cause Zonemaster::Backend to exhibit very surprising behavior.

**Pull request closed without merging**